core/src/main/scala/zio/jdbc/JdbcDecoder.scala

```diff
@@ -581,182 +582,202 @@ trait JdbcDecoderLowPriorityImplicits {
-              // TODO: Timezone
-              val timestamp = resultSet.getTimestamp(columnIndex)
+      case SqlTypes.TIMESTAMP =>
+        val timestamp = resultSet.getTimestamp(columnIndex)
 
-              DynamicValue.Primitive(timestamp.toInstant(), StandardType.InstantType)
+        Some(DynamicValue.Primitive(timestamp.toInstant(), StandardType.InstantType))
 
-            case SqlTypes.TIME_WITH_TIMEZONE =>
-              // TODO: Timezone
-              val time = resultSet.getTime(columnIndex)
+      case SqlTypes.TIMESTAMP_WITH_TIMEZONE =>
+        // TODO: Timezone
+        val timestamp = resultSet.getTimestamp(columnIndex)
 
-              DynamicValue.Primitive(time.toLocalTime(), StandardType.LocalTimeType)
+        Some(DynamicValue.Primitive(timestamp.toInstant(), StandardType.InstantType))
 
-            case SqlTypes.TINYINT =>
-              val short = resultSet.getShort(columnIndex)
+      case SqlTypes.TIME_WITH_TIMEZONE =>
+        // TODO: Timezone
+        val time = resultSet.getTime(columnIndex)
 
-              DynamicValue.Primitive(short, StandardType.ShortType)
+        Some(DynamicValue.Primitive(time.toLocalTime(), StandardType.LocalTimeType))
 
-            case SqlTypes.VARBINARY =>
-              val chunk = getBinary(resultSet.getBinaryStream(columnIndex))
+      case SqlTypes.TINYINT =>
+        val short = resultSet.getShort(columnIndex)
 
-              DynamicValue.Primitive(chunk, StandardType.BinaryType)
+        Some(DynamicValue.Primitive(short, StandardType.ShortType))
 
-            case SqlTypes.VARCHAR =>
-              val string = resultSet.getString(columnIndex)
+      case SqlTypes.VARBINARY =>
+        val chunk = getBinary(resultSet.getBinaryStream(columnIndex))
 
-              DynamicValue.Primitive(string, StandardType.StringType)
+        Some(DynamicValue.Primitive(chunk, StandardType.BinaryType))
 
-            case other =>
-              throw new SQLException(
-                s"Unsupported SQL type ${other} when attempting to decode result set from a ZIO Schema ${schema}"
-              )
-          }
+      case SqlTypes.VARCHAR =>
+        val string = resultSet.getString(columnIndex)
 
-        listMap = listMap.updated(name, value)
+        Some(DynamicValue.Primitive(string, StandardType.StringType))
 
-        columnIndex += 1
-      }
+      case _ =>
+        None
+    }
+  }
 
-      DynamicValue.Record(TypeId.Structural, listMap)
+  private def mapComplexType(resultSet: ResultSet, columnIndex: Int, sqlType: Int): Option[DynamicValue] =
+    sqlType match {
+      case SqlTypes.ARRAY =>
+        val arrayRs   = resultSet.getArray(columnIndex).getResultSet
+        val metaData  = arrayRs.getMetaData
+        var chunkData = Chunk.empty[DynamicValue]
+        while (arrayRs.next())
+          mapPrimitiveType(arrayRs, 2, metaData.getColumnType(2)).foreach { el =>
+            chunkData = chunkData :+ el
+          }
+        Some(DynamicValue.Sequence(chunkData))
```

Comment:
to be more perf/put less pressure on the GC, you'll probably want to use a mutable data structure and then wrap it in a Chunk:

```suggestion
        val data     = mutable.ArrayBuffer.empty[DynamicValue]
        while (arrayRs.next())
          mapPrimitiveType(arrayRs, 2, metaData.getColumnType(2)).foreach { el =>
            data += el
          }
        Some(DynamicValue.Sequence(Chunk.fromArray(data.toArray)))
```